src/utils/format-string.ts

```diff
@@ -1,18 +1,21 @@
 import capitalizeFirstLetter from "../utils/capitalize-first-letter"
 
-const formatString = (str: string): string => {
-  str = str.trim()
-  str = str
-    .replace(/,\s*/g, ", ")
-    .replace(/\.(\s*)/g, ". ")
-    .replace(/\?(\s*)/g, "? ")
-    .replace(/!(\s*)/g, "! ")
+/** Checks if punctuation mark ends without a whitespace at the end.
+ * This represents a paragraph
+ */
+const checkParagraph: RegExp = /[.?!](?!\s)/g
```

Comment:
a paragraph will always end in whitespace unless it is right at the end of the string. This is because `\s` is equivalent to `\r`, `\n`, `\t`, `\f`, and `\v` (check https://regex101.com/ for more info, it is great at checking your regex works). For example:
```
this is a sentence. this is the next sentence.

this should be a new paragraph, but because it is still only separated by the `\n` whitespace `checkParagraph` still considers it as within a single paragraph.
```

Reply:
ah nevermind, i see you use punctuation marks, my bad.